### PR TITLE
Per-receipt hard limit on storage proof size using upper bound estimation

### PR DESCRIPTION
**Pull request by jancionear**
During receipt execution we record all touched nodes from the pre-state trie. Those recorded nodes form the storage proof that is sent to validators, and validators use it to execute the receipts and validate the results.

In https://github.com/near/nearcore/issues/9378 it's stated that in a worst case scenario a single receipt can generate hundreds of megabytes of storage proof. That would cause problems, as it'd cause the `ChunkStateWitness` to also be hundreds of megabytes in size, and there would be problems with sending this much data over the network.

Because of that we need to limit the size of the storage proof. We plan to have two limits:
* per-chunk soft limit - once a chunk has more than X MB of storage proof we stop processing new receipts, and move the remaining ones to the delayed receipt queue. This has been implemented in https://github.com/near/nearcore/pull/10703
* per-receipt hard limit - once a receipt generates more than X MB of storage proof we fail the receipt, similarly to what happens when a receipt goes over the allowed gas limit. This one is implemented in this PR.

Most of the hard-limit code is straightforward - we need to track the size of recorded storage and fail the receipt if it goes over the limit.
But there is one ugly problem: https://github.com/near/nearcore/issues/10890. Because of the way current `TrieUpdate` works we don't record all of the storage proof in real time. There are some corner cases (deleting one of two children of a branch) in which some nodes are not recorded until we do `finalize()` at the end of the chunk. This means that we can't really use `Trie::recorded_storage_size()` to limit the size, as it isn't fully accurate. If we do that, a malicious actor could prepare receipts which seem to have only 1MB of storage proof during execution, but actually record 10MB during `finalize()`.
There is a long discussion in  https://github.com/near/nearcore/issues/10890 along with some possible solution ideas, please read that if you need more context. 

This PR implements Idea 1 from https://github.com/near/nearcore/issues/10890. 
Instead of using `Trie::recorded_storage_size()`  we'll use `Trie::recorded_storage_size_upper_bound()`, which estimates the upper bound of recorded storage size by assuming that every trie removal records additional 2000 bytes:
```rust
    /// Size of the recorded state proof plus some additional size added to cover removals.
    /// An upper-bound estimation of the true recorded size after finalization.
    /// See https://github.com/near/nearcore/issues/10890 and https://github.com/near/nearcore/pull/11000 for details.
    pub fn recorded_storage_size_upper_bound(&self) -> usize {
        // Charge 2000 bytes for every removal
        let removals_size = self.removal_counter.saturating_mul(2000);
        self.recorded_storage_size().saturating_add(removals_size)
    }
```
As long as the upper bound is below the limit we can be sure that the real recorded size is also below the limit.
It's a rough estimation, which often exaggerates the actual recorded size (even by 20+ times), but it could be a good-enough/MVP solution for now. Doing it in a better way would require a lot of refactoring in the Trie code. We're now [moving fast](https://near.zulipchat.com/#narrow/stream/407237-core.2Fstateless-validation/topic/Faster.20decision.20making), so I decided to go with this solution for now.

The upper bound calculation has been added in a previous PR along with the metrics to see if using such a rough estimation is viable: https://github.com/near/nearcore/pull/11000

I set up a mainnet node with shadow validation to gather some data about the size distribution with mainnet traffic: [Metrics link](https://nearinc.grafana.net/d/edbl9ztm5h1q8b/stateless-validation?orgId=1&var-chain_id=mainnet&var-shard_id=All&var-node_id=ci-b20a9aef-mainnet-rpc-europe-west4-01-84346caf&from=1713225600000&to=1713272400000)
image
image

The metrics show that:
* For all receipts both the recorded size and the upper bound estimate are below 2MB
* Overwhelming majority of receipts generate < 50KB of storage proof
* For all chunks the upper bound estimate is below 6MB
* For 99% of chunks the upper bound estimate is below 3MB

Based on this I believe that we can:
* Set the hard per-receipt limit to 4MB. All receipts were below 2MB, but it's good to have a bit of a safety margin here. This is a hard limit, so it might break existing contracts if they turn out to generate more storage proof than the limit.
* Set the soft per-chunk limit to 3MB. 99% of chunks will not be affected by this limit. For the 1% that hit the limit they'll execute fewer receipts, with the rest of the receipts put into the delayed receipt queue. This slightly lowers throughput of a single chunk, but it's not a big slowdown, by ~1%.

Having a 4MB per-receipt hard limit and a 3MB per-chunk soft limit would give us a hard guarantee that for all chunks the total storage proof size is below 7MB.

It's worth noting that gas usage already limits the storage proof size quite effectively. For 98% of chunks the storage proof size is already below 2MB, so the limit isn't really needed for typical mainnet traffic. The limit matters mostly for stopping malicious actors that'd try to DoS the network by generating large storage proofs.

Fixes: https://github.com/near/nearcore/issues/11019

### CODE REVIEW

**Comment by jancionear** [comment]
I still plan to add some tests, but otherwise I'd say the PR is ready for review and we can start discussing the solution. Moving out of draft.

File: core/primitives-core/src/version.rs
Comment (shreyan-gupta) [inline core/primitives-core/src/version.rs:246]: Could you add a quick comment below `ProtocolFeature::StatelessnetShuffleShardAssignmentsForChunkProducers => 84,` saying the next protocol version just implements the hard limit and we should use 86 for the next feature?

File: runtime/near-vm-runner/src/logic/recorded_storage_counter.rs
Comment (shreyan-gupta) [inline runtime/near-vm-runner/src/logic/recorded_storage_counter.rs:11]: I'm wondering if we can push this implementation to the ext.rs layer? We could have this check done for every trie operations like `storage_set`, `storage_get`, `storage_remove`, instead of remembering to have to call it in logic/mod.rs

Comment (pugachAG) [reply runtime/near-vm-runner/src/logic/recorded_storage_counter.rs:11]: The way I see it per-receipt storage counting is very similar to what we have with gas, so it would be nice to have those in the same place to keep it somewhat consistent. Current implementation does exactly that, so I suggest to keep it.

**Review by pugachAG — APPROVED** [review]
LGTM!
One thing to note: in the current implementation previously executed receipts affect storage proof size for the currently executed one (if the value was previously recorded then it won't increase recorded storage size). This potentially could be a problem when a contract is executed successfully after other receipts, but then could fail when executed as a first receipt in the chunk. In practice I don't think it matters and I don't suggest addressing that, I don't believe it is worth the effort and added implementation complexity.
@jancionear please make sure to get an approval from someone from the contract runtime team before merging this

**Comment by Longarithm** [comment]
> previously executed receipts affect storage proof size for the currently executed one (if the value was previously recorded then it won't increase recorded storage size)

Oh, right. Well, I also don't think we should make impl more complex to account for that. As a reference, we already have similar effect for chunk cache.

File: core/primitives-core/src/version.rs
Comment (wacban) [reply core/primitives-core/src/version.rs:246]: Just to make sure I understand, why is there no need to check for it? Is it because it's checked in the runtime params? 

I have to say it is rather confusing to see bumped protocol version without corresponding protocol feature. I would recommend adding it just for clarity or adding a comment somewhere.

File: runtime/near-vm-runner/src/logic/errors.rs
Comment (wacban) [inline runtime/near-vm-runner/src/logic/errors.rs:214]: nit: best to use bytesize for type safety and human friendliness

**Review by nagisa — APPROVED** [review]

File: core/primitives-core/src/version.rs
Comment (jancionear) [reply core/primitives-core/src/version.rs:246]: > Just to make sure I understand, why is there no need to check for it? Is it because it's checked in the runtime params?

Yes, for previous protocol versions the limit is 9999999999 (+inf), so it will never be triggered. The limit becomes 4MB at protocol 85 and it's defined in the runtime params, so there's no need for protocol feature checks.

> I have to say it is rather confusing to see bumped protocol version without corresponding protocol feature. I would recommend adding it just for clarity or adding a comment somewhere.

I will add one :+1:

**Comment by jancionear** [comment]
> I still plan to add some tests

I was able to construct a receipt which produces 24MB of storage proof by reading 24 values, 1MB each. Tested it manually on localnet and it seems that the hard limit works as it should. It rejected the 24MB receipt, but a 3MB receipt went through without any problems. Now I'll try to make it into a proper integration test or something like that (???)

File: runtime/near-vm-runner/src/logic/errors.rs
Comment (jancionear) [reply runtime/near-vm-runner/src/logic/errors.rs:214]: Yeah with a single value it looks much more reasonable:
```
Error: 
   0: Error: An error occurred during a `FunctionCall` Action, parameter is debug message.
      ExecutionError("Size of the recorded trie storage proof has exceeded the allowed limit (4.0 MB)")


```

Comment (jancionear) [reply runtime/near-vm-runner/src/logic/errors.rs:214]: I hope it's okay to add a `bytesize` dependency to `near-vm-runner` :eyes:

**Pull request merged**
